File: crates/koloda-core/src/repo/conversations.rs
use rusqlite::{params, OptionalExtension};
use serde::Deserialize;
use serde_json::Value;

use crate::app::db::Database;
use crate::app::error::AppError;
use crate::app::utility::{deserialize_optional_timestamp, get_current_timestamp};
use crate::domain::conversations::Conversation;

fn get_conversation_row(row: &rusqlite::Row<'_>) -> Result<Conversation, rusqlite::Error> {
    let state_str: String = row.get(1)?;
    let state = serde_json::from_str(&state_str)
        .map_err(|e| rusqlite::Error::FromSqlConversionFailure(1, rusqlite::types::Type::Text, Box::new(e)))?;

    Ok(Conversation {
        id: row.get(0)?,
        state,
        created_at: row.get(2)?,
        updated_at: row.get(3)?,
    })
}

pub fn get_conversation(db: &Database, id: &str) -> Result<Option<Conversation>, AppError> {
    db.with_conn(|conn| {
        conn.query_row(
            r#"
            SELECT id, state, created_at, updated_at
            FROM conversations
            WHERE id = ?1
            LIMIT 1
            "#,
            params![id],
            get_conversation_row,
        )
        .optional()
        .map_err(AppError::from)
    })
}

pub fn get_conversations(db: &Database) -> Result<Vec<Conversation>, AppError> {
    db.with_conn(|conn| {
        let mut stmt = conn.prepare(
            r#"
            SELECT id, state, created_at, updated_at
            FROM conversations
            ORDER BY updated_at DESC, created_at DESC
            "#,
        )?;

        let rows = stmt.query_map([], get_conversation_row)?;
        let mut conversations = Vec::new();
        for row in rows {
            conversations.push(row?);
        }
        Ok(conversations)
    })
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetConversationInput {
    pub id: String,
    pub state: Value,
    #[serde(default, deserialize_with = "deserialize_optional_timestamp")]
    pub updated_at: Option<i64>,
}

pub fn set_conversation(db: &Database, input: SetConversationInput) -> Result<Conversation, AppError> {
    let now = get_current_timestamp()?;

    db.with_conn(|conn| {
        conn.execute(
            r#"
            INSERT INTO conversations (id, state, created_at, updated_at)
            VALUES (?1, ?2, ?3, NULL)
            ON CONFLICT(id) DO UPDATE SET
                state = excluded.state,
                updated_at = COALESCE(?4, ?5)
            "#,
            params![input.id, input.state.to_string(), now, input.updated_at, now],
        )?;
        Ok(())
    })?;

    get_conversation(db, &input.id)?.ok_or_else(|| AppError::new(crate::app::error::error_codes::DB_UPDATE, None))
}

pub fn delete_conversation(db: &Database, id: &str) -> Result<(), AppError> {
    db.with_conn(|conn| {
        conn.execute("DELETE FROM conversations WHERE id = ?1", params![id])?;
        Ok(())
    })
}

File: libs/db-sqlite/src/lib/conversations.ts
import { conversationHasTurns, conversationListItemSchema, conversationRowSchema, throwKnownError } from "@koloda/app";
import type { DeleteConversationData, SetConversationData } from "@koloda/app";
import { CONVERSATION_SELECT } from "./columns";
import type { DB } from "./db";
import { parseRow, parseRowOrNull, parseRows } from "./parse-rows";
import { nowMs } from "./sql";

export async function getConversation(db: DB, id: string) {
  return throwKnownError("db.get", async () => {
    const result = await db.get(`SELECT ${CONVERSATION_SELECT} FROM conversations WHERE id = ? LIMIT 1`, [id]);
    return parseRowOrNull(conversationRowSchema, result);
  });
}

export async function getConversations(db: DB) {
  return throwKnownError("db.get", async () => {
    const result = await db.all(
      `SELECT ${CONVERSATION_SELECT} FROM conversations ORDER BY updated_at DESC, created_at DESC`,
    );
    const rows = parseRows(conversationRowSchema, result);
    return parseRows(
      conversationListItemSchema,
      rows.map((row) => ({
        id: row.id,
        title: row.title,
        createdAt: row.createdAt,
        updatedAt: row.updatedAt,
        hasTurns: conversationHasTurns(row.state),
      })),
    );
  });
}

export async function setConversation(db: DB, { id, state, title, updatedAt }: SetConversationData) {
  return throwKnownError("db.update", async () => {
    const now = nowMs();
    const nextUpdatedAt = updatedAt ? updatedAt.getTime() : now;

    await db.run(
      `INSERT INTO conversations (id, title, state, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?)
       ON CONFLICT(id) DO UPDATE SET
         title = excluded.title,
         state = excluded.state,
         updated_at = excluded.updated_at`,
      [id, title ?? null, JSON.stringify(state), now, nextUpdatedAt],
    );

    const result = await getConversation(db, id);
    return parseRow(conversationRowSchema, result);
  });
}

export async function deleteConversation(db: DB, { id }: DeleteConversationData) {
  return throwKnownError("db.delete", async () => {
    await db.run(`DELETE FROM conversations WHERE id = ?`, [id]);
  });
}
